### PR TITLE
Issue #4394: increase coverage of pitest-checkstyle-api profile to 96%

### PR DESCRIPTION
**Pull request by Nimfadora**
Issue #4394

[pitest-checkstyle-api profile before changes](https://nimfadora.github.io/pr/com.puppycrawl.tools.checkstyle.api/index.html)

increased mutation threshold for pitest-checkstyle-api-filters profile by 8%
current threshold: 96%
execution time: 9:51 min

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java [hunk 1]
```diff
@@ -227,11 +227,12 @@ public int getLineNo() {
             // with initialize(String text)
             resultNo = findLineNo(getFirstChild());
 
-            if (resultNo < 0) {
+            //line number cannot be less than 1
+            if (resultNo == -1) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java]: Why do we add comment here but no where else in this file?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java [hunk 2]
```diff
@@ -49,12 +49,6 @@
  */
 public abstract class AbstractLoader
     extends DefaultHandler {
-    /** Feature that enables loading external DTD when loading XML files. */
-    private static final String LOAD_EXTERNAL_DTD =
-        "http://apache.org/xml/features/nonvalidating/load-external-dtd";
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java]: @romani Please review these removals.

File: src/test/java/com/puppycrawl/tools/checkstyle/api/AbstractCheckTest.java [hunk 3]
```diff
@@ -196,4 +201,42 @@ public void testGetAcceptableTokens() throws Exception {
...
+
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testClearMessages() {
+        final AbstractCheck check = new DummyAbstractCheck();
+
+        check.log(0, "key", "args");
+        final Collection<LocalizedMessage> messages =
+                (Collection<LocalizedMessage>) Whitebox.getInternalState(check, "messages");
+        check.clearMessages();
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/AbstractCheckTest.java]: Can we add an assertion right before this that messages' size isn't 0?

File: src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java [hunk 4]
```diff
@@ -117,10 +124,94 @@ public void testInsertSiblingBetween() throws Exception {
...
+                    throw new IllegalStateException(exception);
+                }
+            }
+        );
+
+        for (Consumer<DetailAST> method : clearBranchTokenTypesMethods) {
+            final BitSet branchTokenTypes = Whitebox.invokeMethod(parent, "getBranchTokenTypes");
+            method.accept(null);
+            final BitSet branchTokenTypes2 = Whitebox.invokeMethod(parent, "getBranchTokenTypes");
+            assertNotSame(branchTokenTypes, branchTokenTypes2);
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java]: We can't do more than `notSame`? It makes this a vague test.

File: src/test/java/com/puppycrawl/tools/checkstyle/api/FullIdentTest.java [hunk 5]
```diff
@@ -37,7 +37,16 @@ public void testToString() {
...
+    public void testNonValidCoordinatesWithNegative() {
+        testNonValidCoordinates(14, 15, "MyTest.MyTestik[15x14]");
+    }
+
+    @Test
+    public void testNonValidCoordinatesWithZero() {
+        testNonValidCoordinates(0, 0, "MyTest.MyTestik[15x14]");
+    }
+
+    private static void testNonValidCoordinates(int columnNo, int lineNo, String expected) {
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/FullIdentTest.java]: Non-test methods shouldn't start with `test`. Rename `test` to `validate`.
Also romani doesn't like assertions outside of test method.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java (hunk 1 above)
Comment (Nimfadora) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java]: I just forgot to remove this one.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java (hunk 2 above)
Comment (Nimfadora) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java]: Those features are enabled by [default](https://xerces.apache.org/xerces2-j/features.html). I moved them to the tests - to always be sure that they are still enabled.

Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java]: @Nimfadora This is not true. Look at blame for these lines and associated issue.
https://github.com/checkstyle/checkstyle/blame/154f120d7b71c30ec64f3ef801fb080a24a21c4a/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java#L52
https://github.com/checkstyle/checkstyle/issues/3605

This removal must be reversed.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java (hunk 1 above)
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java]: Then please remove.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java (hunk 2 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java]: @Nimfadora ,
Unfortunately, this is weird feature request from googlers , please see [commit](https://github.com/checkstyle/checkstyle/commit/636ae05d053e1a2ff59c1060a2736f22ab0f6c1b) and related issue #3605 .
How can we suppress pitest there or any other workaround (I am ok to move that code to special method `addFeauresForVerySecureJavaInstallations` and suppress pitest on this method)?

File: src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java (hunk 4 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java]: I'm not really sure how. BitSets actually are equals each time. By this test, I wanted to ensure that each call to add/set sibling/child methods force bitset to be recalculated.

Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java]: Can we then assert the resulting values are the same?
`NotSame` atleast verifies instance is different.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java (hunk 2 above)
Comment (Nimfadora) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java]: @romani  it's kinda troublesome. I can exclude `addFeauresForVerySecureJavaInstallations` from being mutated, but I cannot suppress mutation of _the call_ to this method. `avoidCallTo` in PIT config didn't help, cause it allows only class names as params.
I can pull this method to the separate class, eg `AbstractLoaderUtils`, and this helps. Is it fine?

**Comment by Nimfadora** [comment]
Build failed with 
`[testClass=com.puppycrawl.tools.checkstyle.api.FileTextTest, name=testLineColumnAfterCopyConstructor(com.puppycrawl.tools.checkstyle.api.FileTextTest)] did not pass without mutation.`
This test definitly passes localy. What have I done wrong?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java (hunk 2 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java]: please try to do this by means of inner class in AbstractLoader, if it does not help ... ok to make separate top level class `FeauresForVerySecureJavaInstallations` in javadoc make it clear that it is result of conflict of pitest and user requirements.

**Comment by rnveach** [comment]
> This test definitly passes localy. What have I done wrong?

First thing that comes to mind is make sure you are up to date with master.
I cannot look into it right now.

**Comment by romani** [comment]
@Nimfadora , on my local for your branch the same on Travis:
```
$ mvn test
......
[ERROR] Failures: 
[ERROR]   FileTextTest.testLineColumnAfterCopyConstructor:92 expected:<44> but was:<46>
```

please recheck that tests are passing before pitest execution as it heavily affect pitest.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java (hunk 2 above)
Comment (Nimfadora) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractLoader.java]: fixed

File: src/test/java/com/puppycrawl/tools/checkstyle/api/AbstractCheckTest.java (hunk 3 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/AbstractCheckTest.java]: fixed

File: src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java (hunk 4 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java]: fixed

File: src/test/java/com/puppycrawl/tools/checkstyle/api/FullIdentTest.java (hunk 5 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/FullIdentTest.java]: fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java (hunk 1 above)
Comment (Nimfadora) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java]: fixed

**Pull request merged**